### ISSUE TITLE
Allow users to specify offset / timestamp to configure starting point for Event Hub

### ISSUE THREAD

**Issue by christopheranderson** [Azure/azure-webjobs-sdk#1240]
Users today cannot control when their Event Hub starts to run from - it generally starts from the very beginning which can mean lots of data which adds latency/wasted compute/etc. Users should be able to specify how far back the Event Hubs trigger starts.

From: https://feedback.azure.com/forums/355860/suggestions/19085314

**Comment by paulbatum** [comment]
I can definitely see how this is useful. Some design work is necessary here - not sure of the best way to expose this.

**Comment by tohling** [comment]
This can be done by setting the `EventProcessorOptions` located in `EventHubConfiguration` at https://github.com/Azure/azure-webjobs-sdk/blob/ffae7c86ea87fd73748186bac2c38c5401b80f68/src/Microsoft.Azure.WebJobs.ServiceBus/EventHubs/EventHubConfiguration.cs#L55

In the current implementation any such settings will be applied Function App wide.  We will need to redesign this so that it will work at the Function level.

**Comment by paulbatum** [comment]
@tohling I don't think this is settable by functions today, which I believe was the intent of this issue (though not explicitly stated)

**Comment by jeffhollan** [comment]
Just adding my 2c that I think the best "experience" comes from the Stream Analytics way of doing this - when creating a stream analytics job you specify either "Start from current time", "Start from beginning", or "Specify start time."  People won't know event identifiers, so letting them specify a DateTime (Which the Event Hub SDK supports I believe) would be ideal. Know that's not necessarily in line with current implementation.

**Comment by jantielens** [comment]
I agree with @jeffhollan, it would be awesome to have something similar to starting a Stream Analytics Job. Although the start options are a bit different compared to Jeff's list:
- Now
- Custom (specific date/time)
- When last stopped (only available when the Job already started/stopped before)

![sastart](https://user-images.githubusercontent.com/9884103/35860280-c72d2e40-0b43-11e8-856f-fa0ca2c7fe6d.JPG)

I believe the current behavior for Azure Functions triggered by an Event Hub is similar to **Now** when the Function starts for the first time. And **When last stopped** when the Functions starts again (after being stopped before). Can someone validate this assumption?

If so, adding **Custom** and **Now** (which is basically Custom set to the current time) would be great, when starting a Function again.

**Comment by otykier** [comment]
Is there a workaround / way to accomplish this, currently? Our scenario is that we want to replay a set of events, so basically directing the consumer group used by our Function to go back in time.

I found [this stackoverflow post](https://stackoverflow.com/questions/47558739/can-i-change-the-azure-function-event-hub-trigger-starting-checkpoint/47562337#47562337) that suggested we can change the checkpoint in blob storage, but it seems a little risky/hacky, and also we can't easily correlate point-in-time to offset values.

Any suggestions?

**Comment by jeffhollan** [comment]
You laid out the way to do this today (modifying checkpoints in blob storage).  Is something on backlog to improve in future though.

**Comment by pvmraghunandan** [comment]
any update on this? We just need to read from the offset if we are creating reader for the first time. Currently it is not possible with event hubs trigger

**Comment by FlavioDiasPs** [comment]
It is been a while, but I still can't find another solution other than manually changing it on blob storage.
Is there a solution current in development for it?

**Comment by paulbatum** [comment]
I think the best approach for adding this feature would be for the eventhub binding extension to add admin endpoints to the runtime that allow for control over checkpoints. This way, we can build several different user experiences (command line, azure portal, etc) that all rely on the same implementation provided by the binding extension. The same approach could be followed for other checkpoint based bindings, such as CosmosDB.

**Comment by jantielens** [comment]
FYI: in the Azure Feedback site, this issue was moved to status "planned" (in November 2018) by @jeffhollan . So I'm optimistic. :) https://feedback.azure.com/forums/355860/suggestions/19085314

**Comment by vullnetyy** [comment]
I need this for testing purposes. I've got an event hub with a 7 day retention policy containing approx 200 million events and I am only interested in the evens happening right now. Even though I used something like:
````csharp
if (eventData.SystemProperties.EnqueuedTimeUtc < DateTime.UtcNow - TimeSpan.FromMinutes(30))
{
    return;
}
````
I will still need to waste 24 hours of compute time (and wait) before I get to the events I'm actually interested in. Is there something I can do to get around this?

**Comment by paulbatum** [comment]
Until we get this feature implemented, you'll need to modify the checkpoints manually. I have not attempted this myself but here is what I would try:

1. Create a simple console app that uses the event processor host nuget package and then use their API to start processing the stream from the point in time you want. Configure this app to use some separate storage account you've made. Run it long enough for it to create checkpoints for each partition in that storage account. They'll look something like this:

![image](https://user-images.githubusercontent.com/10663/63549777-f7797d80-c4e5-11e9-994b-54622ca4a160.png)

2. Stop the function app you have deployed in azure, and then replace the checkpoints in the storage account for your function app with the checkpoints from your local application.

3. Start the function app and see if it picks them up correctly and resumes executing from where you wanted.

Sorry this isn't simpler!

**Comment by ErikAndreas** [comment]
Would also really need a way to tell a hub triggered function where to start (similar to EventPosition in [EventHubClient.CreateReceiver](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.eventhubclient.createreceiver?view=azure-dotnet)). Have a use case where we're really just interested in "current" (as real-time as it gets) data and catching up, like @vullnetyy mentioned is a pain. 

But, I've taken a stab at what @paulbatum suggested and implemented an [AzureIoTHubCheckpointSetter](https://github.com/ErikAndreas/AzureIoTHubCheckpointSetter), PR's welcome!

**Comment by DanielLarsenNZ** [comment]
Another scenario is adding a new Function (with a new consumer group) to an Event Hub that is already in production. We should be able to start the offset from HEAD, Now, Offset or DateTime. Currently the function must loop 7 days of events before getting to head.

How about a new trigger option and two new attribute parameters, e.g.

```json
{
    "eventHubs": {
      "eventProcessorOptions": {
        "initialOffsetStartFrom": "now"
      }
    }    
}  
```

Possible values for `initialOffsetStartFrom` would be "beginning" or "now". Default = "beginning".

And in the Attribute we can be more specific (and configurable).

```csharp
[EventHubTrigger("EventHubName", Connection = "ConnStringName", ConsumerGroup = "ConsumerGroupName", MinDateTime = "Friday, 15 November 2019 12:00:00 AM")]
```

```csharp
[EventHubTrigger("EventHubName", Connection = "ConnStringName", ConsumerGroup = "ConsumerGroupName", MinOffset = 1234567890)]
```

Implementation would be a strategy in a delegate for [InitialOffsetProvider](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.processor.eventprocessoroptions.initialoffsetprovider?view=azure-dotnet).

**Comment by paulbatum** [comment]
I've come up with a solution that might work well for some folks. It uses the [dependency injection feature](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection) to configure the initial offset provider. The example below just uses `DateTime.UtcNow` but you could apply whatever logic you want here.

```csharp

using System;
using Microsoft.Azure.EventHubs;
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Azure.WebJobs.EventHubs;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

[assembly: FunctionsStartup(typeof(EventHubOffsetTest.Startup))]

namespace EventHubOffsetTest
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddOptions<EventHubOptions>()
                .PostConfigure(o => o.EventProcessorOptions.InitialOffsetProvider = (partitionId) => EventPosition.FromEnqueuedTime(DateTime.UtcNow));
        }
    }
}
```

I've done some basic testing and this seems to work as expected. Please be careful as this would apply to all event hub triggered functions in your function app. I am keen to hear from folks on whether this works for you.

Also I should note that if your functions are written in JS/python/etc, it should be possible to use the same approach (but you would need to do it in C# as I have done above), but I don't think we have good documentation for this scenario yet.

**Comment by llevlion** [comment]
Using dependency injection does not seem to work correctly. The offset is still not set for partition "0".
GetInitialOffset() is never called with "0" on startup.

**Comment by DimaTodo89** [comment]
> Using dependency injection does not seem to work correctly. The offset is still not set for partition "0".
> GetInitialOffset() is never called with "0" on startup.

I was able to get the Azure Function to ignore all previous event hub data by using @paulbatum's code above and changing the FromEnqueuedTime(DateTime.UtcNow) to FromEnd().

I also had to consume from a consumer group previously unused by that Azure Function. You could probably achieve this second step by deleting the offset tracking files in the Azure Function storage but I haven't tested that (delete at your own risk).

**Comment by paulbatum** [comment]
@DimaTodo89 thanks for sharing this, not sure how I missed FromEnd() (maybe its a more recent addition to the eventhub SDK?). I updated my code sample to use that.

**Comment by hbensalem** [comment]
@paulbatum  Thanks for sharing this. Is there anyway to set the offset with a previously used consumer group ? Your solution only works for a newly created consumer group.

**Comment by vullnetyy** [comment]
@hbensalem this works with existing consumer groups as well, as long as you delete the offset data stored in azure-webjobs-eventhub in the storage account you connected to your function app. 

Is there a reason you would like to avoid deleting this data? 
Since your function would change the offsets on the consumer group data as soon as it starts running, I find it hard to imagine why you want to keep them. But if you do, could you maybe back them up to another storage account?

**Comment by dwoldo** [comment]
@paulbatum Wonderful mitigation! Thank you. We can feasible set this option via appSettings in such as case we need to start at the end or even a specific period in time.
